commit 24f65983b8f0bc487e9532cd8a8f13483ac66ccb
Author: Manoj Gupta <manojgupta@google.com>
Date:   Thu Dec 20 18:07:41 2018 -0800

    Revert "[debuginfo] generate debug info with asm+.file"

    This reverts commit 3b38b07fc66936279b29230042bb2b0ab9ec6826.

    commit 3b38b07fc66936279b29230042bb2b0ab9ec6826
    Author:     Brian Cain <bcain@codeaurora.org>
    AuthorDate: Tue Aug 28 16:23:39 2018 +0000
    Commit:     Brian Cain <bcain@codeaurora.org>
    CommitDate: Tue Aug 28 16:23:39 2018 +0000

        [debuginfo] generate debug info with asm+.file

	Summary:
        For assembly input files, generate debug info even when the .file
        directive is present, provided it does not include a file-number
        argument.  Fixes PR38695.

	Reviewers: probinson, sidneym

	Subscribers: aprantl, hiraditya, JDevlieghere, llvm-commits

diff --git a/lib/MC/MCParser/AsmParser.cpp b/lib/MC/MCParser/AsmParser.cpp
index a66a4eb29af..e856a7e0287 100644
--- a/llvm/lib/MC/MCParser/AsmParser.cpp
+++ b/llvm/lib/MC/MCParser/AsmParser.cpp
@@ -3369,6 +3369,14 @@ bool AsmParser::parseDirectiveFile(SMLoc DirectiveLoc) {
     }
   }
 
+  // In case there is a -g option as well as debug info from directive .file,
+  // we turn off the -g option, directly use the existing debug info instead.
+  // Also reset any implicit ".file 0" for the assembler source.
+  if (Ctx.getGenDwarfForAssembly()) {
+    Ctx.getMCDwarfLineTable(0).resetRootFile();
+    Ctx.setGenDwarfForAssembly(false);
+  }
+
   if (FileNumber == -1) {
     // Ignore the directive if there is no number and the target doesn't support
     // numberless .file directives. This allows some portability of assembler
@@ -3376,14 +3384,6 @@ bool AsmParser::parseDirectiveFile(SMLoc DirectiveLoc) {
     if (getContext().getAsmInfo()->hasSingleParameterDotFile())
       getStreamer().EmitFileDirective(Filename);
   } else {
-    // In case there is a -g option as well as debug info from directive .file,
-    // we turn off the -g option, directly use the existing debug info instead.
-    // Also reset any implicit ".file 0" for the assembler source.
-    if (Ctx.getGenDwarfForAssembly()) {
-      Ctx.getMCDwarfLineTable(0).resetRootFile();
-      Ctx.setGenDwarfForAssembly(false);
-    }
-
     MD5::MD5Result *CKMem = nullptr;
     if (HasMD5) {
       CKMem = (MD5::MD5Result *)Ctx.allocate(sizeof(MD5::MD5Result), 1);
diff --git a/test/MC/AsmParser/directive_file-3.s b/test/MC/AsmParser/directive_file-3.s
deleted file mode 100644
index c3bdaede270..00000000000
--- a/llvm/test/MC/AsmParser/directive_file-3.s
+++ /dev/null
@@ -1,24 +0,0 @@
-// RUN: llvm-mc -g -triple i386-unknown-unknown  %s | FileCheck -check-prefix=CHECK-DEFAULT %s
-// RUN: llvm-mc -g -triple i386-unknown-unknown %s -filetype=obj | obj2yaml | FileCheck -check-prefix=CHECK-DEBUG %s
-
-// Test for Bug 38695
-// This testcase has a single function and a .file directive
-// without the [file-num] argument.  When compiled with -g,
-// this testcase will not report error, and generate new
-// debug info.
-
-        .file "hello"
-.text
-
-f1:
-        nop
-.size f1, .-f1
-
-// CHECK-DEFAULT: .file "hello"
-
-// CHECK-DEBUG:  Sections:
-// CHECK-DEBUG:  - Name:            .text
-// CHECK-DEBUG:  - Name:            .debug_info
-// CHECK-DEBUG:  - Name:            .rel.debug_info
-// CHECK-DEBUG:    Info:            .debug_info
-// CHECK-DEBUG:  Symbols:
